src/lib.rs

```diff
@@ -1,90 +1,221 @@
+// Note that this will be linked in from the Julia process, we do not try
+// to link it while building this Rust lib.
+#[cfg(feature = "julia")]
+extern "C" {
+    fn jl_adopt_thread() -> i32;
+    fn jl_gc_safe_enter() -> i32;
+    fn jl_gc_disable_finalizers_internal() -> c_void;
 }
 
-struct IcebergScanInternal {
-    table: Option<iceberg::table::Table>,
-    columns: Option<Vec<String>>,
-    stream: Option<Mutex<futures::stream::BoxStream<'static, Result<RecordBatch, iceberg::Error>>>>,
+// Simple response type for operations that only need success/failure status
+#[repr(C)]
+pub struct IcebergResponse {
+    result: CResult,
+    error_message: *mut c_char,
+    context: *const Context,
 }
 
-// Global Tokio runtime using OnceLock for thread safety
-// TODO: Might want to share tokio runtime between here and object_store_ffi.jl, e.g.,
-// by passing object store in and using its runtime.
-static RUNTIME: OnceLock<Runtime> = OnceLock::new();
+unsafe impl Send for IcebergResponse {}
 
-fn get_runtime() -> &'static Runtime {
-    RUNTIME.get_or_init(|| Runtime::new().expect("Failed to create Tokio runtime"))
-}
+impl RawResponse for IcebergResponse {
+    type Payload = ();
+
+    fn result_mut(&mut self) -> &mut CResult {
+        &mut self.result
+    }
+
+    fn context_mut(&mut self) -> &mut *const Context {
+        &mut self.context
+    }
+
+    fn error_message_mut(&mut self) -> &mut *mut c_char {
+        &mut self.error_message
+    }
 
-// Thread-local error storage
-thread_local! {
-    static LAST_ERROR: std::cell::RefCell<Option<String>> = std::cell::RefCell::new(None);
+    fn set_payload(&mut self, _payload: Option<Self::Payload>) {
+        // No payload for simple response
+    }
 }
 
-fn set_error(error: String) {
-    LAST_ERROR.with(|e| {
-        *e.borrow_mut() = Some(error);
-    });
+// Callback types for Julia integration
+type PanicCallback = unsafe extern "C" fn() -> i32;
+
+// Simple config for iceberg - only what we need
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub struct IcebergStaticConfig {
+    n_threads: usize,
 }
 
-fn clear_error() {
-    LAST_ERROR.with(|e| {
-        *e.borrow_mut() = None;
-    });
+impl Default for IcebergStaticConfig {
+    fn default() -> Self {
+        IcebergStaticConfig {
+            n_threads: 0, // 0 means use tokio's default
+        }
+    }
 }
 
-// C API structures
+// Direct structures - no opaque wrappers
 #[repr(C)]
 pub struct IcebergTable {
-    _private: [u8; 0], // Opaque type for C
+    pub table: Table,
 }
 
 #[repr(C)]
 pub struct IcebergScan {
-    _private: [u8; 0], // Opaque type for C
+    pub builder: Option<TableScanBuilder<'static>>,
+    pub scan: Option<TableScan>,
+}
+
+unsafe impl Send for IcebergScan {}
+
+// Stream wrapper for FFI - using async mutex to avoid blocking calls
+#[repr(C)]
+pub struct IcebergArrowStream {
+    pub stream:
+        AsyncMutex<futures::stream::BoxStream<'static, Result<RecordBatch, iceberg::Error>>>,
```

Comment:
I thought about this a bit more yesterday, and I think we don't need mutex at all?

We could just declare it unsafe to access the stream from multiple threads and that's it.

Just like it's not safe to access any other object in Julia from multiple threads without acquiring lock, e.g. dictionary (and for those that are safe, there are separate structs, like ConcurrentDict).